core/lib/zksync_core/src/db_pruner/mod.rs

```diff
@@ -0,0 +1,501 @@
+
+        let next_miniblock_to_prune = transaction
+            .blocks_dal()
+            .get_miniblock_range_of_l1_batch(next_l1_batch_to_prune)
+            .await?
+            .unwrap()
+            .1;
+        transaction
+            .pruning_dal()
+            .soft_prune_batches_range(next_l1_batch_to_prune, next_miniblock_to_prune)
+            .await?;
+
+        transaction.commit().await?;
+
+        let latency = latency.observe();
+        tracing::info!(
+            "Soft pruned db l1_batches up to {} and miniblocks up to {}, operation took {:?}",
+            next_l1_batch_to_prune,
+            next_miniblock_to_prune,
+            latency
+        );
+
+        return Ok(true);
+    }
+
+    async fn hard_prune(&self, pool: &ConnectionPool) -> anyhow::Result<()> {
+        let latency = METRICS.pruning_chunk_duration[&MetricPruneType::Hard].start();
+
+        let mut storage = pool.access_storage().await?;
+        let mut transaction = storage.start_transaction().await?;
+
+        let current_pruning_info = transaction.pruning_dal().get_pruning_info().await?;
+        transaction
+            .pruning_dal()
+            .hard_prune_batches_range(
+                current_pruning_info.last_soft_pruned_l1_batch.unwrap(),
+                current_pruning_info.last_soft_pruned_miniblock.unwrap(),
+            )
+            .await?;
+
+        transaction.commit().await?;
+
+        let latency = latency.observe();
+        tracing::info!(
+            "Hard pruned db l1_batches up to {} and miniblocks up to {}, operation took {:?}",
+            current_pruning_info.last_soft_pruned_l1_batch.unwrap(),
+            current_pruning_info.last_soft_pruned_miniblock.unwrap(),
+            latency
+        );
+
+        Ok(())
+    }
+
+    pub async fn run_single_iteration(&self, pool: &ConnectionPool) -> anyhow::Result<()> {
+        let mut storage = pool.access_storage().await?;
+
+        let mut current_pruning_info = storage.pruning_dal().get_pruning_info().await?;
+
+        if current_pruning_info.last_soft_pruned_l1_batch
+            == current_pruning_info.last_hard_pruned_l1_batch
+        {
+            let pruning_done = self.soft_prune(&pool).await?;
+            if !pruning_done {
+                return Ok(());
+            }
+
+            tokio::time::sleep(self.config.soft_and_hard_pruning_time_delta).await;
+        }
+
+        self.hard_prune(&pool).await?;
+
+        Ok(())
+    }
+    pub async fn run_in_loop(
+        self,
+        pool: ConnectionPool,
+        stop_receiver: watch::Receiver<bool>,
+    ) -> anyhow::Result<()> {
+        loop {
+            if *stop_receiver.borrow() {
+                tracing::info!("Stop signal received, shutting down DbPruner");
+            }
+            let _ = self.update_l1_batches_metric(&pool).await;
+            // as this component is not really mission-critical, all errors are generally ignored
+            let _ = self.run_single_iteration(&pool).await;
```

Comment:
Bikeshedding: I don't think ignoring errors is a reasonable option in this case; e.g., if we have logical errors or there's something terribly wrong with the DB, it makes more sense to stop the node.

Reply:
I'd argue that if we make a bug and pruning has stopped for whatever reason, I'd prefer to learn about if from alerts about pruning being off rather than crashlooping node. It's only critical that we'd know about each such situation

Reply:
I added logging and we've had a discussion about this